RFC/src/RFC-0010_CodeStructure.md

```diff
@@ -0,0 +1,117 @@
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).
+
+## Disclaimer
+
+The purpose of this document and its content is for information purposes only and may be subject to change or update
+without notice.
+
+This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
+community. The release of this document is intended solely for review and discussion by the community regarding the
+technological merits of the potential system outlined herein.
+
+## Goals
+
+This RFC describes and explains the Tari codebase layout.
+
+## Related RFCs
+
+None.
+
+## Description
+
+The code follows a domain-driven design layout ([DDD]), with top-level folders falling into infrastructure, domain and
+application layers.
+
+### Infrastructure layer
+
+This layer provides a set of crates which have general infrastructural utility. The rest of the Tari codebase can make use
+of these crates to obtain persistence, communication and cryptographic services. The infrastructure layer doesn't know
+anything about blockchains, transactions, or digital assets.
+
+We recommend that code in this layer generalises infrastructure services behind abstraction layers as much as is
+reasonable, so that specific implementations can be swapped out with relative ease.
+
+### Domain layer
+
+The Domain layer houses the Tari "business logic". All protocol-related concepts and procedures are defined and
+implemented here.
+
+This entails that any and all terms defined in the [Glossary] will have a software implementation here, and only here.
+They can be _used_ in the Application layer, but must be implemented in the Domain layer.
+
+The domain layer can make use of crates in the infrastructure layer to achieve its goals.
+
+### Application layer
+
+Applications build on top of the domain layer to produce the executable software that is deployed as part of the Tari
+network.
+
+As an example, the following base layer applications may be developed as part of the Tari protocol release:
+
+* A standalone miner (tari_miner)
+* A pool miner (tari_pool_miner)
+* A CLI wallet for the Tari cryptocurrency (cli_wallet)
+* A base node executable (tari_basenode)
+* A REST API server for the base node
+
+### Code layout
+
+The top-level source code directories in this repository reflect the respective [DDD] layers; except that there are two
+domain layer directories, corresponding to the two network layers that make up the Tari network.
+
+1. `infrastructure` folder contains code that is not Tari-specific. It holds the following crates:
+    - `comms`: The networking and messaging subsystem
+    - `crypto`: All cryptographic services, including a Curve25519 implementation
+   - `storage`: Data persistence services, including an LMDB persistence implementation
```

Comment:
Add `consensus` as another container in `infrastructure`.

Reply:
Can you give more reasoning here please. Consensus sounds more like a domain-level construct.

Reply:
Consensus can contain different consensus mechanisms and primitives, to be used as required in the next layers

Reply:
I doubt it's needed for base layer; for DAN, we can add it when it becomes clearly required.